src/main/java/com/icthh/xm/ms/configuration/service/TenantAliasService.java

```diff
@@ -49,12 +52,44 @@ public List<Configuration> processConfiguration(Configuration configuration,
         try {
             TenantAliasTree tenantAliasTree = mapper.readValue(configuration.getContent(), TenantAliasTree.class);
             tenantAliasTree.init();
+
+            TenantAliasTree oldTenantAliasTree = this.tenantAliasTree;
             // safe publication
             this.tenantAliasTree = tenantAliasTree;
-            tenantAliasTree.getTenants().keySet().forEach(configurationService::refreshTenantConfigurations);
+
+            var oldTenants = oldTenantAliasTree.getTenants();
+            var newTenants = tenantAliasTree.getTenants();
+            var allTenants = new HashSet<String>();
+            allTenants.addAll(oldTenants.keySet());
+            allTenants.addAll(newTenants.keySet());
+
+            allTenants.stream().filter(it -> isTenantChanged(oldTenants, newTenants, it))
+                    .map(newTenants::get)
+                    .map(this::getParentKey)
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .distinct()
+                    .forEach(memoryConfigStorage::reprocess);
+
+            allTenants.stream().filter(it -> isTenantChanged(oldTenants, newTenants, it))
+                    .distinct()
+                    .forEach(configurationService::refreshTenantConfigurations);
+
```

Comment:
What do you think of this approach?
```suggestion
            allTenants.stream().filter(it -> isTenantChanged(oldTenants, newTenants, it))
                    .map(newTenants::get)
                    .map(this::getParentKey)
                    .filter(Optional::isPresent)
                    .map(Optional::get)
                    .distinct()
                    .peek(memoryConfigStorage::reprocess)
                    .filter(it -> isTenantChanged(oldTenants, newTenants, it))
                    .distinct()
                    .forEach(configurationService::refreshTenantConfigurations);


```

Reply:
this is not equals change

Reply:
refreshTenantConfigurations by tenants

process by parents